### PR TITLE
Improve codebase

### PR DESCRIPTION
**Pull request by nyamsprod**
- Adding missing return type for the ValidatorInterface (should be boolean)
- Using DateTimeImmutable instead of DateTime

### CODE REVIEW

**Comment by DragonBe** [comment]
Hey Ignace,

I reviewed your changes and it all makes sense what your goal is, thank you for that. But when we're running our tests things start to break. Can you review your changes and see why the tests fail after your changes?

```
PHPUnit 6.2.2 by Sebastian Bergmann and contributors.

F

Time: 190 ms, Memory: 6.00MB

There was 1 failure:

1) DragonBe\Test\Vies\CheckVatResponseTest::testCanCreateResponseAtConstruct with data set #0 (true)
Failed asserting that two variables reference the same object.

/Users/dragonbe/workspace/dragonbe/vies/tests/Vies/CheckVatResponseTest.php:84

FAILURES!
Tests: 1, Assertions: 3, Failures: 1.
```

Another thing I wanted to ask you is that when you provide a return type in the class method, maybe you can also specify the type of the argument? This way we solve two issues at one single go. Can you have another look at this please?

screenshot 2017-06-15 16 10 57

Thank you for your contribution, this means a lot to me and all users of this tool!

File: src/Vies/CheckVatResponse.php [hunk 1]
```diff
@@ -34,7 +34,7 @@ class CheckVatResponse
      */
     protected $vatNumber;
     /**
-     * @var \DateTime The date of the request
+     * @var \DateTimeImmutable The date of the request
```
Comment (DragonBe) [inline src/Vies/CheckVatResponse.php]: Not sure if [DateTimeImmutable](https://secure.php.net/datetimeimmutable) is necessary here. Can you clarify why you have decided to make it immutable?

File: src/Vies/CheckVatResponse.php [hunk 2]
```diff
@@ -109,28 +109,27 @@ public function getVatNumber(): string
     /**
      * Sets the date- and timestamp when the VIES service response was created
      *
-     * @param \DateTime $requestDate
+     * @param \DateTimeInterface $requestDate
```
Comment (DragonBe) [inline src/Vies/CheckVatResponse.php]: Good call to make the parameter an interface! 👍

**Comment by nyamsprod** [comment]
@DragonBe thanks for the review. yep I'll try to work on the PR next week I have my hand full right now :'( 

Maybe I should split the PR in two 
- one for the DateTimeImmutable
- the other for the return type ?

What do you think ?

File: src/Vies/CheckVatResponse.php [hunk 3]
```diff
@@ -109,28 +109,27 @@ public function getVatNumber(): string
...
      */
-    public function setRequestDate(\DateTime $requestDate): CheckVatResponse
+    public function setRequestDate(\DateTimeInterface $requestDate): CheckVatResponse
     {
-        if (! $requestDate instanceof \DateTime) {
-            $date = substr($requestDate, 0, 10);
-            $timezone = substr($requestDate, -6);
-            $requestDate = new \DateTime($date, \DateTime::createFromFormat('O', $timezone)->getTimezone());
+        if ($requestDate instanceof \DateTime) {
+            $requestDate = \DateTimeImmutable::createFromMutable($requestDate);
```
Comment (DragonBe) [inline src/Vies/CheckVatResponse.php]: Is this a necessary step? Aren't both `DateTime` and `DateTimeImmutable` implement the `DateTimeInterface`?

File: src/Vies/CheckVatResponse.php [hunk 4]
```diff
@@ -109,28 +109,27 @@ public function getVatNumber(): string
...
         }
+
         $this->requestDate = $requestDate;
         return $this;
     }
     /**
      * Retrieves the date- and timestamp the VIES service response was created
      *
-     * @return \DateTime
+     * @return \DateTimeImmutable
```
Comment (DragonBe) [inline src/Vies/CheckVatResponse.php]: Wouldn't `DateTimeInterface` be a cleaner return type?

File: src/Vies/CheckVatResponse.php [hunk 5]
```diff
@@ -109,28 +109,27 @@ public function getVatNumber(): string
...
      */
-    public function getRequestDate(): \DateTime
+    public function getRequestDate(): \DateTimeImmutable
```
Comment (DragonBe) [inline src/Vies/CheckVatResponse.php]: Wouldn't `DateTimeInterface` be a cleaner return type?

File: src/Vies/HeartBeat.php [hunk 6]
```diff
@@ -100,12 +100,7 @@ public function setPort(int $port): heartbeat
      */
     public function isAlive(): bool
     {
-
-        $status = $this->connect($this->getHost(), $this->getPort());
-        if (false === $status) {
-            return false;
-        }
-        return true;
+        return false !== $this->connect($this->getHost(), $this->getPort());
```
Comment (DragonBe) [inline src/Vies/HeartBeat.php]: Clean improvement, thank you 👍

File: src/Vies/HeartBeat.php [hunk 7]
```diff
@@ -120,8 +115,7 @@ private function connect(string $host, int $port): bool
     {
         $status = HeartBeat::$testingServiceIsUp;
         if (false === HeartBeat::$testingEnabled) {
-            $resource = fsockopen($host, $port);
-            return (false !== $resource);
+            return false !== fsockopen($host, $port);
```
Comment (DragonBe) [inline src/Vies/HeartBeat.php]: Clean improvement, thank you 👍

File: src/Vies/Validator/ValidatorAT.php [hunk 8]
```diff
@@ -34,7 +34,7 @@ class ValidatorAT extends ValidatorAbstract
      * @param string $vatNumber
      * @return bool
      */
-    public function validate($vatNumber)
+    public function validate($vatNumber): bool
```
Comment (DragonBe) [inline src/Vies/Validator/ValidatorAT.php]: Maybe type-hint the argument in the same go

**Comment by DragonBe** [comment]
@nyamsprod Sure thing, take your time. I'm already happy you've taken time to offer this PR, but I understand you have other engagements.

The idea to split up the PR's is a good idea, please continue down that path.

Again, many thanks for contributing!

Michelangelo

File: src/Vies/CheckVatResponse.php (hunk 3 above)
Comment (nyamsprod) [inline src/Vies/CheckVatResponse.php]: Yes they are but by normalizing to DateTimeImmutable you make sure that CheckVatResponse will always return the immutable version.

File: src/Vies/CheckVatResponse.php (hunk 4 above)
Comment (nyamsprod) [inline src/Vies/CheckVatResponse.php]: Depends on how the DateTime object is used. If it is changed outside the CheckVatResponse the if a DateTime is returned it means that someone can change CheckVatResponse state without touching the object. That the reason I believe switching to DateTimeImmutable is important. On the otherhand the fact that you accept any type of DateTimeInterface implementing object means that a Carbon or a Chronos object could be accepted. So the accepted range is wide but the return object is strict.

File: src/Vies/HeartBeat.php (hunk 7 above)
Comment (nyamsprod) [inline src/Vies/HeartBeat.php]: I thought I would do that everywhere but I refrained myself for doing it  I could finish this on in the return type PR if you want

File: src/Vies/CheckVatResponse.php (hunk 3 above)
Comment (DragonBe) [inline src/Vies/CheckVatResponse.php]: Good idea, I just need to make sure that it doesn't break current implementations of VIES.

File: src/Vies/CheckVatResponse.php (hunk 4 above)
Comment (DragonBe) [inline src/Vies/CheckVatResponse.php]: Interesting chain of thoughts. Once the immutable type is defined, with this addition we make sure that the return type is always immutable.

File: src/Vies/HeartBeat.php (hunk 7 above)
Comment (DragonBe) [inline src/Vies/HeartBeat.php]: maybe do this in a separate PR for cleaning up the code

**Comment by nyamsprod** [comment]
@DragonBe please refer to PR #29 where I've put the PHP7.1 improvements. When it is merge I'll work on the DateTime PR.

**Comment by DragonBe** [comment]
Can I close this PR? PR #29 is the leading one to follow, right?

**Comment by nyamsprod** [comment]
@DragonBe yes you can
